File: analytics/src/test/java/io/appmetrica/analytics/impl/component/RegularDispatcherComponentHandleReportTest.java
package io.appmetrica.analytics.impl.component;

import android.content.Context;
import io.appmetrica.analytics.impl.CounterReport;
import io.appmetrica.analytics.impl.InternalEvents;
import io.appmetrica.analytics.impl.component.clients.ClientUnit;
import io.appmetrica.analytics.impl.component.clients.ComponentUnitFactory;
import io.appmetrica.analytics.impl.request.StartupArgumentsTest;
import io.appmetrica.analytics.impl.request.StartupRequestConfig;
import io.appmetrica.analytics.impl.startup.StartupCenter;
import io.appmetrica.analytics.impl.startup.StartupUnit;
import io.appmetrica.analytics.testutils.CommonTest;
import java.util.ArrayList;
import java.util.Collection;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.ParameterizedRobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(ParameterizedRobolectricTestRunner.class)
public class RegularDispatcherComponentHandleReportTest extends CommonTest {

    @Mock
    private ComponentId mComponentId;
    private StartupRequestConfig.Arguments startupArguments = StartupArgumentsTest.empty();
    private CommonArguments.ReporterArguments reportArguments = CommonArgumentsTestUtils.emptyReporterArguments();
    private CommonArguments mClientConfiguration;
    @Mock
    private ComponentUnitFactory<ComponentUnit> mComponentUnitFactory;
    @Mock
    private ComponentUnit mReportingComponent;
    @Mock
    private ComponentLifecycleManager<ClientUnit> mLifecycleManager;
    @Mock
    private ReporterArgumentsHolder mReporterArgumentsHolder;

    private RegularDispatcherComponent<ComponentUnit> mRegularDispatcherComponent;
    private CounterReport mCounterReport;
    private Context mContext;

    private int mEventType;

    public RegularDispatcherComponentHandleReportTest(int eventType) {
        mEventType = eventType;
    }

    @ParameterizedRobolectricTestRunner.Parameters(name = "[{index}]Event with type = {0} handle as diagnostic ? {1}")
    public static Collection<Object[]> data() {
        Collection<Object[]> data = new ArrayList<Object[]>();
        for (InternalEvents eventType : InternalEvents.values()) {
            data.add(new Object[]{eventType.getTypeId()});
        }

        return data;
    }

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.openMocks(this);
        when(mReporterArgumentsHolder.getArguments()).thenReturn(reportArguments);
        mContext = RuntimeEnvironment.getApplication();
        when(mComponentUnitFactory.createComponentUnit(same(mContext), same(mComponentId), any(CommonArguments.ReporterArguments.class), any(StartupUnit.class)))
                .thenReturn(mReportingComponent);
        mClientConfiguration = new CommonArguments(startupArguments, reportArguments, null);
        StartupCenter startupCenter = mock(StartupCenter.class);
        doReturn(mock(StartupUnit.class)).when(startupCenter).getOrCreateStartupUnit(any(Context.class), any(ComponentId.class), any(StartupRequestConfig.Arguments.class));
        mRegularDispatcherComponent = new RegularDispatcherComponent<ComponentUnit>(
                mContext,
                mComponentId,
                mClientConfiguration,
                mReporterArgumentsHolder,
                mComponentUnitFactory,
                mLifecycleManager,
                startupCenter
        );

        mCounterReport = new CounterReport();
        mCounterReport.setType(mEventType);
    }

    @Test
    public void testHandleReport() {
        mRegularDispatcherComponent.handleReport(mCounterReport, CommonArgumentsTestUtils.createMockedArguments());
        verify(mReportingComponent, times(1)).handleReport(mCounterReport);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/component/RegularDispatcherComponentHandleReportTest.kt
package io.appmetrica.analytics.impl.component

import android.content.Context
import io.appmetrica.analytics.impl.CounterReport
import io.appmetrica.analytics.impl.InternalEvents
import io.appmetrica.analytics.impl.component.clients.ClientUnit
import io.appmetrica.analytics.impl.component.clients.ComponentUnitFactory
import io.appmetrica.analytics.impl.request.StartupArgumentsTest
import io.appmetrica.analytics.impl.startup.StartupCenter
import io.appmetrica.analytics.impl.startup.StartupUnit
import io.appmetrica.analytics.testutils.CommonTest
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.same
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.robolectric.ParameterizedRobolectricTestRunner

@RunWith(ParameterizedRobolectricTestRunner::class)
class RegularDispatcherComponentHandleReportTest(
    eventType: Int,
    private val shouldUpdateConfig: Boolean
) : CommonTest() {

    private val context: Context = mock()
    private val componentId: ComponentId = mock()
    private val startupArguments = StartupArgumentsTest.empty()
    private val reportArguments = CommonArgumentsTestUtils.emptyReporterArguments()
    private val reportingComponent: ComponentUnit = mock()
    private val startupUnit: StartupUnit = mock()

    private val startupCenter: StartupCenter = mock {
        on { getOrCreateStartupUnit(same(context), any(), any()) } doReturn startupUnit
    }

    private val clientConfiguration: CommonArguments = CommonArguments(startupArguments, reportArguments, null)

    private val componentUnitFactory: ComponentUnitFactory<ComponentUnit> = mock {
        on { createComponentUnit(same(context), same(componentId), any(), any()) } doReturn reportingComponent
    }

    private val lifecycleManager: ComponentLifecycleManager<ClientUnit?> = mock()
    private val counterReport: CounterReport = CounterReport(null, eventType)

    private val regularDispatcherComponent: RegularDispatcherComponent<ComponentUnit> by setUp {
        RegularDispatcherComponent(
            context,
            componentId,
            clientConfiguration,
            componentUnitFactory,
            lifecycleManager,
            startupCenter
        )
    }

    @Test
    fun handleReport() {
        regularDispatcherComponent.handleReport(counterReport, CommonArgumentsTestUtils.createMockedArguments())
        verify(reportingComponent).handleReport(counterReport)
        verify(reportingComponent, times(if (shouldUpdateConfig) 1 else 0)).updateSdkConfig(reportArguments)
    }

    companion object {

        @JvmStatic
        @ParameterizedRobolectricTestRunner.Parameters(name = "[{index}]Event with type = {0} updateConfig ? {1}")
        fun data(): Collection<Array<Any>> = InternalEvents.values().map {
            arrayOf(
                it.typeId,
                it != InternalEvents.EVENT_TYPE_UPDATE_FOREGROUND_TIME &&
                    it != InternalEvents.EVENT_TYPE_EXCEPTION_UNHANDLED_FROM_FILE &&
                    it != InternalEvents.EVENT_TYPE_PREV_SESSION_EXCEPTION_UNHANDLED_FROM_FILE &&
                    it != InternalEvents.EVENT_TYPE_PREV_SESSION_NATIVE_CRASH_PROTOBUF &&
                    it != InternalEvents.EVENT_TYPE_CURRENT_SESSION_NATIVE_CRASH_PROTOBUF
            )
        }
    }
}
